cmd/notation/internal/plugin/plugin.go

```diff
@@ -0,0 +1,138 @@
+
+package plugin
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/notaryproject/notation-go/dir"
+	"github.com/notaryproject/notation-go/plugin"
+	"github.com/notaryproject/notation-go/plugin/proto"
+	notationerrors "github.com/notaryproject/notation/cmd/notation/internal/errors"
+	"github.com/notaryproject/notation/internal/osutil"
+	"github.com/opencontainers/go-digest"
+)
+
+// PluginSourceType is an enum for plugin source
+type PluginSourceType string
+
+const (
+	// TypeFile means plugin source is file
+	TypeFile PluginSourceType = "file"
+
+	// TypeURL means plugin source is URL
+	TypeURL PluginSourceType = "url"
+
+	// TypeUnknown means unknown plugin source
+	TypeUnknown PluginSourceType = "unknown"
+)
+
+const (
+	// TypeZip means plugin file is zip
+	TypeZip = "application/zip"
+
+	// TypeGzip means plugin file is gzip
+	TypeGzip = "application/x-gzip"
+)
+
+// CheckPluginExistence returns true if a plugin already exists
+func CheckPluginExistence(ctx context.Context, pluginName string) (bool, error) {
+	mgr := plugin.NewCLIManager(dir.PluginFS())
+	_, err := mgr.Get(ctx, pluginName)
+	if err != nil {
+		if errors.Is(err, os.ErrNotExist) {
+			return false, nil
+		}
+		return false, err
+	}
+	return true, nil
+}
+
+// ValidateCheckSum returns nil if SHA256 of file at path equals to checkSum.
+func ValidateCheckSum(path string, checkSum string) error {
+	r, err := os.Open(path)
+	if err != nil {
+		return err
+	}
+	defer r.Close()
+	dgst, err := digest.FromReader(r)
+	if err != nil {
+		return err
+	}
+	enc := dgst.Encoded()
+	if enc != checkSum {
+		return fmt.Errorf("plugin checksum does not match user input. Expecting %s", checkSum)
+	}
+	return nil
+}
+
+// ValidatePluginMetadata validates plugin metadata given plugin name and path,
+// returns the plugin version on success
+func ValidatePluginMetadata(ctx context.Context, pluginName, path string) (string, error) {
+	plugin, err := plugin.NewCLIPlugin(ctx, pluginName, path)
+	if err != nil {
+		return "", err
+	}
+	metadata, err := plugin.GetMetadata(ctx, &proto.GetMetadataRequest{})
+	if err != nil {
+		return "", err
+	}
+	return metadata.Version, nil
+}
+
+// DownloadPluginFromURL downloads plugin file from url to a tmp directory
+// it returns the tmp file path of the downloaded file
+func DownloadPluginFromURL(ctx context.Context, url, tmpDir string) (string, error) {
+	// Create the file
+	tmpFilePath := filepath.Join(tmpDir, "notationPluginTmp")
+	out, err := os.OpenFile(tmpFilePath, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0600)
+	if err != nil {
+		return "", err
+	}
+	defer out.Close()
+	// Get the data
+	resp, err := http.Get(url)
+	if err != nil {
+		return "", err
+	}
+	defer resp.Body.Close()
+	// Check server response
+	if resp.StatusCode != http.StatusOK {
+		return "", fmt.Errorf("bad status: %s", resp.Status)
+	}
+	// Writer the body to file
+	_, err = io.Copy(out, resp.Body)
```

Comment:
There is no limit on the `resp.Body`. Therefore, the code is vulnerable to excessive downloading attack. A malicious server can blow up your disk.

Reply:
The size of the response could vary greatly depending on the size of the plugin. Do we have a suggested size limit here?

Reply:
256 MB should be large enough as it is 10x of a compressed plugin.

Reply:
> 256 MB should be large enough as it is 10x of a compressed plugin.

I was about to propose 64MB, but sure, 256MB sounds good to me as well.